scripts/rest_service.py

```diff
@@ -0,0 +1,303 @@
+
+    def get_msg(self) -> Any:
+        try:
+            return self._queue.get_nowait()
+        except Exception:
+            pass
+
+    def send(self, msg: Any) -> None:
+        self._queue.put(msg)
+
+
+# ------------------------------------------------------------------------------
+#
+class Service(object):
+
+    def __init__(self, app: FastAPI) -> None:
+
+        self._clients: Dict[str, _Client] = dict()
+        self._deserialize = psij.Import()
+        self._log = logging.getLogger('psij')
+        self._cnt: int = 0
+
+        # ----------------------------------------------------------------------
+        # websocket endpoint at which client can regoister for state updates
+        @app.websocket("/ws/{cid}")
+        async def ws_endpoint(ws: WebSocket, cid: str) -> None:
+
+            await ws.accept()
+
+            client = self._clients.get(cid)
+            if not client:
+                self._log.error("refuse ws for %s" % cid)
+                raise ValueError("unknown client cid %s" % cid)
+
+            self._log.info("accept ws for %s" % cid)
+            try:
+                # keep this async task alive as long as there are messages to
+                # send and the websocket is alive
+                while True:
+                    msg = client.get_msg()
+                    if msg:
+                        await ws.send_json(msg)
+                    else:
+                        await asyncio.sleep(0.1)
+
+            except WebSocketDisconnect:
+                self._log.info("dropped ws for %s" % cid)
+        # ----------------------------------------------------------------------
+
+    def _status_callback(self, cid: str, job: psij.Job, status: psij.JobStatus) -> None:
+
+        client = self._clients.get(cid)
+        if not client:
+            print("unknown client cid %s" % cid)
+            return
+
+        msg = {'jobid': job.id,
+               'time': status.time,
+               'message': status.message,
+               'state': str(status.state),
+               'metadata': status.metadata,
+               'exit_code': status.exit_code}
+        self._log.debug('cb: %s: %s', cid, msg)
+        client.send(msg)
+
+    def _request_register(self, name: str, url: Optional[str] = None) -> str:
+        '''
+        parameters:
+            name:str: name of psij executor to use
+            url:str: optional URL to be passed to backend executor
+
+        returns:
+            str: unique ID identifying the registered client
+        '''
+
+        # register new client
+        cid = 'client.%04d' % self._cnt
+        self._cnt += 1
+
+        # create executor
+        jex = psij.JobExecutor.get_instance(name=name, url=url)
+
+        # register state callback for this cid
+        cb = functools.partial(self._status_callback, cid)
+        jex.set_job_status_callback(cb)
+
+        # store client information
+        self._clients[cid] = _Client(jex)
```

Comment:
I can't find a place where clients are removed from `_clients`. A client-facing call named `register` would suggest that a complementary `unregister` might be there. If that's the solution going forward, I'd be wary, since clients can disappear at any time and there are no guarantees that proper cleanup is done.

I would generally favor either cleaning based on job state (if a state if final, unregister, but that might pose some problems with polling) or finding a stateless solution.

Reply:
Fair, an `unregister` would make sense, even if just for symmetry reasons, I'll add it (also to the ZMQ implementation where the same applies).  

As for cleaning idle clients after a while: we should then also allow to reconnect, e.g. support registering the same client ID again as a way to reconnect.  But it also might be sufficient for now to leave that problem to be resolved on the client side: if a call with an outdated client ID fails because the client was cleaned out after a while, the client should still be able to re-register for a new ID and then re-attach all it's job IDs.  This could be implemented in the client's REST executor.  That assumes that the backend executor can perform that reattach call, which I think is true for all executors we have (and is in fact the very purpose of that reattach anyway).  Would that work for you?

Reply:
Yes, something along those lines. Would probably need some thinking and design work to smooth out.